File: 39-CombinationSum/39-CombinationSum.py
# Last updated: 24/07/2026, 23:54:53
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []
        res = []
        def helper(i:int,sum:int):
            if i >= len(candidates) or sum<0:
                return
            if sum == 0:
                result.append(list(res))
                return
            res.append(candidates[i])
            helper(i,sum-candidates[i])
            res.pop()
            helper(i+1,sum)
            return
        helper(0,target)
        return result
        
        

File: 39-CombinationSum/39-CombinationSum.js
// Last updated: 24/07/2026, 23:55:00
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function (candidates, targets) {
    let result = [];
    let helper = (i, res, target) => {
        if (i >= candidates.length || target < 0) return;
        if (target == 0) {
            result.push([...res])
            return;
        }
        res.push(candidates[i])
        helper(i, res, target - candidates[i]);
        res.pop();
        helper(i + 1, res, target);
        return;
    }
    helper(0, [], targets);
    return result;
};